video-rvfc/request-video-frame-callback-repeating.html WPT fails

jernoble opened the pull request:
#### 60289bcb5d45540417e8c1b9cb08be64dd3a9e52
<pre>
video-rvfc/request-video-frame-callback-repeating.html WPT fails
<a href="https://bugs.webkit.org/show_bug.cgi?id=282585">https://bugs.webkit.org/show_bug.cgi?id=282585</a>
<a href="https://rdar.apple.com/139251513">rdar://139251513</a>

Reviewed by Jean-Yves Avenard.

Implement both `expectedDisplayTime` and `mediaTime` properties of
VideoFrameMetadata in MediaPlayerPrivateWebM. To do so, modify
VideoMediaSampleRenderer::copyDisplayedPixelBuffer() to return
a combination of a pixel buffer and that buffer&apos;s presentation time.

* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.h:
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
(WebCore::MediaPlayerPrivateWebM::updateLastPixelBuffer):
(WebCore::MediaPlayerPrivateWebM::checkNewVideoFrameMetadata):
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.h:
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
(WebCore::VideoMediaSampleRenderer::copyDisplayedPixelBuffer):

Canonical link: <a href="https://commits.webkit.org/286451@main">https://commits.webkit.org/286451@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fc332aef61c7b271cd0a79eeb3d9fd5cf064c1d5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/74989 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/54419 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/27806 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/79422 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/26228 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/77106 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/63555 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/2204 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/58869 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17137 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/78056 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/49027 "Found 1 new test failure: media/media-vp8-hiddenframes.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/64421 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39256 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/46336 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/21922 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/24560 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/67473 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/22264 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/80907 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/2307 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/1397 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/67129 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/2456 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/64440 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66429 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16731 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10369 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/8528 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/2272 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/5060 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/2300 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/3221 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/2307 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
```diff
@@ -1741,7 +1743,7 @@ static bool isCopyDisplayedPixelBufferAvailable()
 }
 
-void MediaPlayerPrivateWebM::checkNewVideoFrameMetadata(CMTime currentTime)
+void MediaPlayerPrivateWebM::checkNewVideoFrameMetadata(CMTime currentCMTime)
```
Can we move this to use MediaTime ?

jernoble replied:
We could, but it would just move the place where we need to convert from CMTime -> MediaTime to the caller.

jyavenard commented on Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
```diff
@@ -1750,11 +1752,20 @@ static bool isCopyDisplayedPixelBufferAvailable()
+    if (!presentationTime.isValid())
+        presentationTime = currentTime;
+
+    auto displayTime = MonotonicTime::now().secondsSinceEpoch().seconds() - (currentTime - presentationTime).toDouble();
```
you can now create a MediaTime from `Seconds` `MediaTime::createWithSeconds` 

I was a bit confused here by this calculation. 
I think ` MonotonicTime::now().secondsSinceEpoch().seconds() + (presentationTime - currentTime)` is rather more telling.

as currentTime - presentationTime is always going to be negative.

jyavenard commented on Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
```diff
@@ -1750,11 +1752,20 @@ static bool isCopyDisplayedPixelBufferAvailable()
     metadata.height = m_naturalSize.height();
     metadata.presentedFrames = ++m_sampleCount;
-    metadata.presentationTime = PAL::CMTimeGetSeconds(currentTime);
+    metadata.presentationTime = displayTime;
```
this seems wrong.

A [DOMHighResTimeStamp](https://developer.mozilla.org/en-US/docs/Web/API/DOMHighResTimeStamp) representing the time when the browser submitted the frame for composition.

so it's when we pushed the frame to be decoded, not when we expect it to be displayed  which is what [expectedDisplayTime](https://developer.mozilla.org/en-US/docs/Web/API/HTMLVideoElement/requestVideoFrameCallback#expecteddisplaytime) is

jernoble replied:
We don't have that information, and this behavior matches what we do in other players.

jyavenard commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -414,16 +414,19 @@ - (void)resetUpcomingSampleBufferPresentationTimeExpectations;
-        return buffer;
+    if (!m_decompressionSession) {
+        if (auto buffer = adoptCF([renderer() copyDisplayedPixelBuffer]))
+            return { WTFMove(buffer), MediaTime::invalidTime() };
```
shouldn't we return the wall time instead?

jernoble replied:
No. This is returning the presentationTimeStamp in the media timeline. We don't get that information from -copyDisplayedPixelBuffer.

jyavenard replied:
isn't the currentTime a better estimation than nothing ?

jyavenard commented on Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
```diff
@@ -1750,11 +1752,20 @@ static bool isCopyDisplayedPixelBufferAvailable()
     if (!updateLastPixelBuffer())
```
`updateLastPixelBuffer()` will continue to return nil if the window isn't composited as it calls directly `[AVSampleBufferDisplayLayer copyDisplayedPixelBuffer]` ; shouldn't this be changed to call the VideoMediaSampleRenderer.mm version instead?

jernoble committed:
video-rvfc/request-video-frame-callback-repeating.html WPT fails
https://bugs.webkit.org/show_bug.cgi?id=282585
rdar://139251513

Reviewed by Jean-Yves Avenard.

Implement both `expectedDisplayTime` and `mediaTime` properties of
VideoFrameMetadata in MediaPlayerPrivateWebM. To do so, modify
VideoMediaSampleRenderer::copyDisplayedPixelBuffer() to return
a combination of a pixel buffer and that buffer's presentation time.

* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.h:
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
(WebCore::MediaPlayerPrivateWebM::updateLastPixelBuffer):
(WebCore::MediaPlayerPrivateWebM::checkNewVideoFrameMetadata):
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.h:
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
(WebCore::VideoMediaSampleRenderer::copyDisplayedPixelBuffer):

Canonical link: https://commits.webkit.org/286451@main
